File: src/main/java/de/fluxparticle/animation/point/DynamicPoint2D.java
package de.fluxparticle.animation.point;

import javafx.animation.Interpolatable;
import javafx.beans.binding.DoubleExpression;
import javafx.beans.property.SimpleDoubleProperty;

public class DynamicPoint2D implements Interpolatable<DynamicPoint2D> {

	private final DoubleExpression x;

	private final DoubleExpression y;

	public DynamicPoint2D(DoubleExpression x, DoubleExpression y) {
		this.x = x;
		this.y = y;
	}

	public DynamicPoint2D(double x, double y) {
		this.x = new SimpleDoubleProperty(x);
		this.y = new SimpleDoubleProperty(y);
	}

	@Override
	public DynamicPoint2D interpolate(DynamicPoint2D endValue, double t) {
        DoubleExpression x = this.xProperty().multiply(1.0 - t).add(endValue.xProperty().multiply(t));
        DoubleExpression y = this.yProperty().multiply(1.0 - t).add(endValue.yProperty().multiply(t));
		return new DynamicPoint2D(x, y);
	}

	public double getX() {
		return x.get();
	}

	public DoubleExpression xProperty() {
		return x;
	}

	public double getY() {
		return y.get();
	}

	public DoubleExpression yProperty() {
		return y;
	}
}

File: src/main/kotlin/de/fluxparticle/animation/point/DynamicPoint2D.kt
package de.fluxparticle.animation.point

import javafx.animation.Interpolatable
import javafx.beans.binding.DoubleExpression
import javafx.beans.property.SimpleDoubleProperty

class DynamicPoint2D : Interpolatable<DynamicPoint2D> {

    private val x: DoubleExpression

    private val y: DoubleExpression

    constructor(x: DoubleExpression, y: DoubleExpression) {
        this.x = x
        this.y = y
    }

    constructor(x: Double, y: Double) {
        this.x = SimpleDoubleProperty(x)
        this.y = SimpleDoubleProperty(y)
    }

    override fun interpolate(endValue: DynamicPoint2D, t: Double): DynamicPoint2D {
        val x = this.xProperty().multiply(1.0 - t).add(endValue.xProperty().multiply(t))
        val y = this.yProperty().multiply(1.0 - t).add(endValue.yProperty().multiply(t))
        return DynamicPoint2D(x, y)
    }

    fun getX(): Double {
        return x.get()
    }

    fun xProperty(): DoubleExpression {
        return x
    }

    fun getY(): Double {
        return y.get()
    }

    fun yProperty(): DoubleExpression {
        return y
    }
}
